### ISSUE TITLE
Tutorial on how to prevent silent failures?

### ISSUE THREAD

**Issue by CodeOtter** [mochajs/mocha#2992]
Mocha appears to be extremely prone to failing silently.  I'm not sure if this is because of the child spawning process (which, btw, makes it Mocha also very hard to debug), but when using mocha in a codebase that uses Babel and an extreme reliance on promises, syntax errors or thrown exceptions cause mocha to end with no warning whatsoever.  It's wildly frustrating.

Are there any tutorials on ways to increase Mocha's verbosity when such errors occur?

**Comment by ScottFreeCode** [comment]
Mocha shouldn't be silently failing in the first place (assuming your code isn't suppressing errors in some way or calling `process.exit`, either of which could be impossible for Mocha -- or any other library -- to do anything about); can you provide a more specific example (e.g. code I can copy-paste into a new empty project to reproduce one of each kind of these silent failures)? We don't get many complaints about this (and we have so many users we get complaints even about non-obvious bugs and obvious user errors), so I have a feeling it's not just a generic problem but something unusual about the codebase in question -- but that's not necessarily important *if* we can isolate any particular issue.

**Comment by kevinburke** [comment]
here is one that I ran into: https://github.com/airbrake/node-airbrake/issues/151

another one from today was an Express route that threw a synchronous exception due to a syntax error in the route body. we had configured an error handler, but the error handler was set up to run *before* the routes were declared, so it wasn't actually logging anything when the server errored. (while not really Mocha's fault, it's another issue where it caused tests to fail for a mysterious reason)

previously, I have run into problems with the Waterline ORM throwing objects instead of Errors, older versions of Mocha did not print anything in this case - it would say "Error:" and then not print anything about the thrown object. This is fixed in more recent versions of Mocha and Waterline.

highly recommend registering handlers for uncaught exceptions and uncaught promises, and immediately logging the error/crashing the process in each case.
